bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/SharedValueMapValueInjector.java

```diff
@@ -21,106 +21,234 @@
+        // it is possible for getSharedPropertiesPath to return null for a resource if its resource type is invalid
+        // i.e., String.EMPTY, sling:nonexisting, nt:unstructured, or /var/absolute/path/not/in/resolver/search/path
+        if (sharedPropertiesPath == null) {
+            // if SHARED is requested in this case, return ValueMap.EMPTY
+            if (valueType == SharedComponentProperties.ValueTypes.SHARED) {
+                return ValueMap.EMPTY;
+            } else { // otherwise, pass null for sharedProperties and merge with global and resource
+                return sharedComponentProperties.mergeProperties(globalVmBound, null, resource);
+            }
+        }
+
+        // it is also possible for the shared properties path to differ from the path in bindings for the same
+        // reason that shared properties page path may differ (a wrapped request for a child resource of a
+        // different resource type)
+        final ValueMap sharedVmBound;
+        if (sharedPropertiesPath.equals(bindings.get(SharedComponentProperties.SHARED_PROPERTIES_PATH))) {
+            sharedVmBound = (ValueMap) bindings.get(SharedComponentProperties.SHARED_PROPERTIES);
+        } else {
+            sharedVmBound = Optional.ofNullable(resource.getResourceResolver().getResource(sharedPropertiesPath))
+                    .map(Resource::getValueMap).orElse(ValueMap.EMPTY);
+        }
+
+        // now that shared properties have been retrieved either return them directly if SHARED is requested
+        if (valueType == SharedComponentProperties.ValueTypes.SHARED) {
+            return Optional.ofNullable(sharedVmBound).orElse(ValueMap.EMPTY);
+        } else { // or return everything merged
+            return sharedComponentProperties.mergeProperties(globalVmBound, sharedVmBound, resource);
+        }
+    }
+
+    private SharedComponentProperties.ValueTypes getValueType(final AnnotatedElement element) {
+        return element.getAnnotation(SharedValueMapValue.class).type();
+    }
+
+    private SlingBindings getBindings(Object adaptable) {
+        if (adaptable instanceof ServletRequest) {
+            ServletRequest request = (ServletRequest) adaptable;
+            return (SlingBindings) request.getAttribute(SlingBindings.class.getName());
+        } else {
+            return null;
+        }
     }
 
     /**
      * Get shared properties ValueMap the current resource.
      */
-    protected ValueMap getSharedProperties(String rootPagePath, Resource resource) {
-        String sharedPropsPath = rootPagePath + "/" + JcrConstants.JCR_CONTENT + "/" + SharedComponentProperties.NN_SHARED_COMPONENT_PROPERTIES + "/" + resource.getResourceType();
-        Resource sharedPropsResource = resource.getResourceResolver().getResource(sharedPropsPath);
-        return sharedPropsResource != null ? sharedPropsResource.getValueMap() : ValueMap.EMPTY;
+    protected ValueMap getSharedProperties(Resource resource) {
```

Comment:
Should `getSharedProperties` and `getGlobalProperties` be consolidated in the new `SharedComponentPropertiesImpl` since we also need it there for `getSharedValueMapResourceAdapter`?

Reply:
done.